### PR TITLE
merge PV-Tuning into AQLM main 

### PR DESCRIPTION
**Pull request by justheuristic**
Required

- [x] all fixes from https://github.com/galqiwi/AQLM/pull/1
- [ ] __update readme to use new finetuning__
- [x] explain model type conversion in readme; explain how to run old finetuning
- [x] update conversion to include args.pt for future exporting
- [x] support legacy finetuning mode as part of finetune_fsdp, deprecate finetune_legacy
- [x] black/isort

Optional
- [x] memory-efficient loss for large vocabularies
- [x] better offloading variants for PV finetuning

Validation
- [x] full training cycle from initial calibration to P and PV to conversion, using the new code, for some 7B model
- [x] verify quality vs main
- [x] verify P/PV it/sec does not degrade (...much)
- [x] verify that 70B works, maybe not to convergence

### CODE REVIEW

**Comment by justheuristic** [comment]
README - still need to copy from pv-tuning branch

File: src/datautils.py [hunk 1]
```diff
@@ -248,3 +250,76 @@ def get_loaders(
...
+    result = {
+        k: [t[i : i + block_size] for i in range(0, total_length, block_size)] for k, t in concatenated_examples.items()
+    }
+    if add_labels:
+        result["labels"] = result["input_ids"].copy()
+    return result
+
+
+@torch.no_grad()
+def evaluate_perplexity(
```
Comment (Vahe1994) [inline src/datautils.py:288]: why is this migrated here?

**Review by Vahe1994 — APPROVED** [review]

File: finetune.py [hunk 2]
```diff
@@ -15,182 +48,24 @@
...
-        "--dataset",
-        type=str,
-        help="Dataset name [c4, pajama] or path to data where to extract calibration data from.",
-    )
     parser.add_argument(
-        "--nsamples",
-        type=int,
-        default=1024,
-        help="number of samples",
+        "--monkeypatch_old_pickle",
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply finetune.py:66]: This option fixes pickle errors that we get when swapping between environments - the errors that @Godofnothing and @galqiwi  periodically bump into . So, unfortunately, it is needed. However, if you insist, I can drop it from the main branch.

File: convert_legacy_model_format.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+    args = parser.parse_args()
+    assert args.p_finetuned_state_dict or args.pv_fsdp_dir, "either one of those must be specified"
+    print(f"{args.p_finetuned_state_dict=}, {args.pv_fsdp_dir=}")
+    assert (args.p_finetuned_state_dict is not None) != (args.pv_fsdp_dir is not None)
+
+    args.load_dtype = getattr(torch, args.load_dtype) if args.load_dtype != "auto" else "auto"
+    args.code_dtype = getattr(torch, args.code_dtype) if args.code_dtype is not None else None
+
+    if not args.monkeypatch_old_pickle:
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply convert_legacy_model_format.py:167]: True. As a small consolation, using the prmary way with `--monkeypatch_old_pickle` will **always** work, it's just not the most efficient way.

File: convert_legacy_model_format.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        )
+    elif args.pv_fsdp_dir:
+        quantized_model = load_quantized_model_from_fdsp_checkpoint(
+            args.base_model,
+            args.pv_fsdp_dir,
+            dtype=args.load_dtype,
+            trust_remote_code=args.trust_remote_code,
+        )
+
+    for module in quantized_model.modules():
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply convert_legacy_model_format.py:191]: It is a bit unclear. Please elaborate.

File: src/beam_search_l2.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    flat_unscaled_reference = flat_unscaled_reference.flatten(2, 3).flatten(0, 1)
+    flat_prev_codes = prev_codes.flatten(0, -2)
+    flat_codebooks = codebooks.flatten(-2, -1).detach()
+    dim_order = list(range(num_codebooks))
+    if dim_rng is not None:
+        dim_rng.shuffle(dim_order)
+
+    def _update_flat_codes(_flat_reference, _flat_codes):
+        """update _flat_codes [num_groups, num_codebooks] to approximate _flat_reference [num_groups, group_size]"""
+        if num_codebooks == 1 and beam_size == 1 and stochastic_rounding_tau == 0 and not force_update:
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply src/beam_search_l2.py:88]: Yep, this is a closure. Would it help if we somehow comment on this fact?

File: src/beam_search_l2.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+    stochastic_rounding_tau: float = 0.0,
+    chunk_size_bytes: int = 2**32,
+    dim_rng: Optional[random.Random] = None,
+    force_update: bool = False,
+    max_update_fraction: float = 1.0,
+    code_selection_temperature: float = 0,
+    trust_ratio: Optional[float] = None,
+) -> torch.Tensor:
+    """
+    Update codes using beam search to minimize L2 error in code values (regardless of activations)
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply src/beam_search_l2.py:29]: This is a close relative of beam search from AQ. It solves the same problem as AQ, but also incorporates the tricks from LSQ that can be used on GPU.

Namely, this beam search starts from previous solution, whereas AQ beam search starts from scratch
https://github.com/arbabenko/Quantizations/blob/master/aqCoding.py#L37

File: convert_legacy_model_format.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
+"""
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply convert_legacy_model_format.py:1]: This file allows backward compatibility with models quantized with older version

File: convert_legacy_model_format.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (justheuristic) [reply convert_legacy_model_format.py:191]: Resolution: this is a temporary backwards compatibility patch for users tht train with previous main branch and finetune afterwards. It should be deleted in the nearest PR after 2-4 weeks

File: src/beam_search_l2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (justheuristic) [reply src/beam_search_l2.py:88]: TODO yozh will post a commentary explaining this

File: src/datautils.py (hunk 1 above)
Comment (justheuristic) [reply src/datautils.py:288]: Resolution : keep it here

File: src/beam_search_l2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Vahe1994) [reply src/beam_search_l2.py:88]: The function is using variables from outer scope.
